### DISCUSSION TITLE
Hibernate Sequence Generator 결정 관련 설정

### DISCUSSION THREAD

**Discussion by her0807** [woowacourse-teams/2022-ternoko#364] [Announcements]
## 기본 키 전략 방식으로 @Id, @ generatedValue 를 했다. 

이 속성을 사용해서 기본 키를 생성할 때, 고려할 점을 알게 되어 공유하고 고민하고 싶다. 

## 우선 기존에 사용하고 있는 `IDENTITY` 전략 
기본키 생성을 DB에게 위임한다. (MySQL의 경우는 AUTO INCREMENT)
즉 특정 DB 벤더에 의존적이다.
Id가 null일 경우 해당 객체의 Id를 DB의 AUTO_INCREMENT를 가져와 할당한다. 

따라서 JPA 를 사용해서 객체를 생성할 때 쓰기 지연을 활용하지 못하고 바로 Insert 쿼리가 나가게 된다. 

## Table 을 사용해서 시퀀스 흉내내는 ID 할당 방식
ID 값만 저장하는 TABLE 을 만들고 해당 테이블에서 최신 ID 를 가져와서 데이터에 할당하기 때문에 
특정 벤더에 의존적이지 않다.

시퀀스 테이블을 만들어서 데이터베이스 시퀀스를 흉내낼 Id를 할당한다. 
이 전략을 사용할 시, jpa ddl auto 설정이 되어 있지 않았다면 해당 시퀀스 테이블 생성이 선행이 되어야한다. 
@TableGenerator(테이블 생성 어노테이션)과 함께 사용할 수 있다.
만약 jpa ddl auto 설정이 되어있다면 hibernate가 시퀀스 테이블을 생성한다.

```java

@GeneratedValue(strategy=GenerationType.TABLE, generator = "MEMBER_SEQ_GENERATOR")
@TableGenerator(
name="MEMBER_SEQ_GENERATOR",
table="MY_SEQUENCE", //시퀀스 생성용 테이블 이름
pkColumnName="sequence_name", //MY_SEQUENCE 테이블에 생성할 필드이름(시퀀스네임)
pkColumnValue="MEMBER_SEQ", //SEQ_NAME이라고 지은 칼럼명에 들어가는 값.(키로 사용할 값)
allocationSize=50
)
private Long id;

// ...

```
# 논의 사항
현재 로컬 환경은 H2 DB 를 사용하고, 개발, 운영은 Mysql 을 사용하여 
DataCleaner 생성시 id auto increment 가 초기화되는 동작 방식이 달라서 
2시간 가량 트러블 슈팅을 한  경험이 있습니다. 이 경험을 통해서 DB 호환에 대한 문제의식을 느꼈어요. 

현재 Id 생성 방식이 mysql DB 에 종속적으로 이뤄지고 있는 것 같아서 팀원 분들이
Table 전략도 있다는 점을 참고하여 현행 유지를 할지, table 로 변경해보면 어떨지 , 더 나은 대안이 있는지 
얘기를 나누고 싶습니다.

**Comment by dongho108** [comment]
기본키 식별자 매핑방법에 대한 좋은 의견 감사합니다. 💯

1. `따라서 JPA 를 사용해서 객체를 생성할 때 쓰기 지연을 활용하지 못하고 바로 Insert 쿼리가 나가게 된다.`

식별자가 없는 경우에만 insert쿼리 한번으로 id를 가져오는데 그 다음부턴 쓰기지연을 사용할 수 있습니다.
mysql을 사용하기 때문에 어쩔 수 없는 부분이고 저는 한번정도는 성능에 크게 영향을 안미친다고 생각해요

2. `현재 로컬 환경은 H2 DB 를 사용하고, 개발, 운영은 Mysql 을 사용하여
DataCleaner 생성시 id auto increment 가 초기화되는 동작 방식이 달라서
2시간 가량 트러블 슈팅을 한 경험이 있습니다. 이 경험을 통해서 DB 호환에 대한 문제의식을 느꼈어요.`

저도 h2에서는 auto increment가 없다는 것을 몰라서 몇시간 같이 트러블슈팅을 해봤는데요.
test에서 사용하는 db때문에 로컬, 개발, 운영 DB의 id 매핑전략까진 바꿀필요가 없다고 생각해요
h2가 mysql의 차이점을 정리해서 참고하는 방식정도가 적당하다고 생각합니다. 😊

**Comment by soominsohn** [comment]
기본키 생성전략에 대해 좋은 Discussion 생성 고맙습니다.

제 의견은
IDENTITY 현행 유지입니다!
- DBMS를 MySQL 에서 다른 벤더로 바꿀 예정이 있는가?
- 테스트환경에서 겪은 트러블슈팅의 문제점 때문에 프로덕션 코드 환경을 바꿔야 하는가?

만약 TABLE 전략으로 바꾸었을 때 장점이 뚜렷하면 바꾸면 좋겠지만, 현재로써는 jpa ddl-auto 설정도 validate이기 때문에 바꾸었을 때 사이드이펙트가 더 클 것 같습니다. 

위와 같은 이유로 현행 유지에 한표 합니다.

**Comment by Juhyung990122** [comment]
저도 현행유지에 한표 하겠습니다~! 
위에 말씀주신대로 처음 한번만 겪을 것 같은 문제이기도 하고 
프로덕션 매핑전략까지 바꾸는것보단 이대로 가는 게 좋을 것 같아요.
